hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/ReadMapP.java

```diff
@@ -0,0 +1,885 @@
+            HazelcastInstance client = newHazelcastClient(asClientConfig(clientXml));
+            try {
+                HazelcastClientProxy clientProxy = (HazelcastClientProxy) client;
+                ClientPartitionService partitionService = clientProxy.client.getClientPartitionService();
+                return partitionService.getPartitionCount();
+            } finally {
+                client.shutdown();
+            }
+        }
+
+        @Override
+        @Nonnull
+        public Function<Address, ProcessorSupplier> get(@Nonnull List<Address> addresses) {
+            List<Integer> partitions = IntStream.rangeClosed(0, remotePartitionCount - 1).boxed().collect(toList());
+            return address -> new RemoteProcessorSupplier<>(
+                    clientXml,
+                    reader,
+                    roundRobinSubList(partitions, addresses.indexOf(address), addresses.size())
+            );
+        }
+    }
+
+    private static class RemoteProcessorSupplier<R, E> implements ProcessorSupplier {
+
+        static final long serialVersionUID = 1L;
+
+        private final String clientXml;
+        private final Reader<R, E> reader;
+        private final List<Integer> ownedPartitions;
+
+        private transient HazelcastInstance client;
+        private transient MigrationWatcher migrationWatcher;
+
+        RemoteProcessorSupplier(
+                @Nonnull String clientXml,
+                @Nonnull Reader<R, E> reader,
+                @Nonnull List<Integer> ownedPartitions
+        ) {
+            this.clientXml = clientXml;
+            this.reader = reader;
+            this.ownedPartitions = ownedPartitions;
+        }
+
+        @Override
+        public void init(@Nonnull Context context) {
+            client = newHazelcastClient(asClientConfig(clientXml));
+            reader.init(client);
+            migrationWatcher = new MigrationWatcher(client);
+        }
+
+        @Override
+        public void close(Throwable error) {
+            if (migrationWatcher != null) {
+                migrationWatcher.deregister();
+            }
+            if (client != null) {
+                client.shutdown();
+            }
+        }
+
+        @Override
+        @Nonnull
+        public List<Processor> get(int count) {
+            return processorToPartitions(count, ownedPartitions)
+                    .values().stream()
+                    .map(this::processorForPartitions)
+                    .collect(toList());
+        }
+
+        private Processor processorForPartitions(List<Integer> partitions) {
+            if (partitions.isEmpty()) {
+                return Processors.noopP().get();
+            } else {
+                return new ReadMapP<>(
+                        reader,
+                        partitions,
+                        migrationWatcher.createWatcher()
+                );
+            }
+        }
+    }
+
+    protected interface Reader<R, E> extends Serializable {
```

Comment:
why do you need both interface and abstract class here when interface is never implemented directly?